src/index.ts

```diff
@@ -345,17 +353,123 @@ em.addListener('SANITIZED_CONFIG', function () {
   // TODO: Add function here
 })
 
-// EVENT EMITTER - RESPONSE_RECEIVED
-interface ResponseReceived {
+// EVENT EMITTER - PROBE_RESPONSE_RECEIVED
+interface ProbeResponseReceived {
   alerts: string[]
   response: AxiosResponseWithExtraData
 }
 
-// RESPONSE_RECEIVED - VALIDATE RESPONSE
-em.on(RESPONSE_RECEIVED, function (data: ResponseReceived) {
+// PROBE_RESPONSE_RECEIVED - VALIDATE RESPONSE
+em.on(PROBE_RESPONSE_RECEIVED, function (data: ProbeResponseReceived) {
   const res = validateResponse(data.alerts, data.response)
 
-  em.emit(RESPONSE_VALIDATED, res)
+  em.emit(PROBE_RESPONSE_VALIDATED, res)
+})
+
+// EVENT EMITTER - PROBE_STATUS_PROCESSED
+interface ProbeStatusProcessed {
+  probe: Probe
+  statuses?: StatusDetails[]
+  notifications?: Notification[]
+  validatedResponseStatuses: ValidateResponseStatus[]
+  totalRequests: number
+}
+
+interface ProbeSendNotification extends Omit<ProbeStatusProcessed, 'statuses'> {
+  index: number
+  status?: StatusDetails
+}
+
+interface ProbeSaveLogToDatabase
+  extends Omit<
+    ProbeStatusProcessed,
+    'statuses' | 'totalRequests' | 'validatedResponseStatuses'
+  > {
+  index: number
+  status?: StatusDetails
+}
+
+const probeSendNotification = async (data: ProbeSendNotification) => {
+  const {
+    index,
+    probe,
+    status,
+    notifications,
+    totalRequests,
+    validatedResponseStatuses,
+  } = data
+
+  const statusString = status?.isDown ? 'DOWN' : 'UP'
+  const url = probe.requests[totalRequests - 1].url ?? ''
+
+  if ((notifications?.length ?? 0) > 0) {
+    await sendAlerts({
+      url: url,
+      status: statusString,
+      probeId: probe.id,
+      probeName: probe.name,
+      incidentThreshold: probe.incidentThreshold,
+      notifications: notifications ?? [],
+      validation: validatedResponseStatuses[index],
+    })
+  }
+}
+
+const probeSaveLogToDatabase = async (data: ProbeSaveLogToDatabase) => {
+  const { index, probe, status, notifications } = data
+
+  const type =
+    status?.state === 'UP_TRUE_EQUALS_THRESHOLD'
+      ? 'NOTIFY-INCIDENT'
+      : 'NOTIFY-RECOVER'
+
+  if ((notifications?.length ?? 0) > 0) {
+    Promise.all(
+      notifications?.map((notification) => {
+        const alertMsg = probe.alerts[index]
+
+        return notificationLog({
+          type,
+          probe,
+          alertMsg,
+          notification,
+        })
+      })!
+    )
+  }
+}
+
+// PROBE_STATUS_PROCESSED
+em.on(PROBE_STATUS_PROCESSED, (data: ProbeStatusProcessed) => {
+  const {
+    probe,
+    statuses,
+    notifications,
+    totalRequests,
+    validatedResponseStatuses,
+  } = data
+
+  statuses
+    ?.filter((status) => status.shouldSendNotification)
+    ?.forEach(async (status, index) => {
+      await probeSendNotification({
+        index,
+        probe,
+        status,
+        notifications,
+        totalRequests,
+        validatedResponseStatuses,
+      })
+
+      await probeSaveLogToDatabase({
```

Comment:
I don't think we need to await the `probeSendNotification` and `probeSaveLogToDatabase` function. Just catch the error and log it. It will make these functions executed asynchronously.

```javascript
      probeSendNotification({
        index,
        probe,
        status,
        notifications,
        totalRequests,
        validatedResponseStatuses,
      }).catch(err => log.error(err.message));

      probeSaveLogToDatabase({
        index,
        probe,
        status,
        notifications,
      }).catch(err => log.error(err.message));
```

Reply:
Hmm, make sense. Thanks!